x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useEffect, useRef, useCallback } from 'react';
+import { isEqual } from 'lodash';
+import {
+  Map,
+  MapboxOptions,
+  NavigationControl,
+  Popup,
+  GeoJSONSourceOptions
+} from 'mapbox-gl';
+import { GeoJsonProperties } from 'geojson';
+
+interface ChoroplethDataElement {
+  key: string;
+  value: number;
+  [property: string]: any;
+}
+
+interface Props {
+  style?: React.CSSProperties;
+  initialGeojsonSource: NonNullable<GeoJSONSourceOptions['data']>;
+  geojsonKeyProperty: string;
+  data: ChoroplethDataElement[];
+  renderTooltip: (props: {
+    geojsonProperties: NonNullable<mapboxgl.MapboxGeoJSONFeature['properties']>;
+    data?: ChoroplethDataElement;
+  }) => React.ReactElement | null;
+  initialMapboxOptions?: Partial<MapboxOptions>;
+}
+
+const CHOROPLETH_LAYER_ID = 'choropleth_layer';
+const CHOROPLETH_POLYGONS_SOURCE_ID = 'choropleth_polygons';
+
+const linearScale = (x: number, range = { min: 0, max: 1 }) =>
+  (range.max - range.min) * x + range.min;
+const quadradicScale = (x: number, range = { min: 0, max: 1 }) =>
+  4 * (range.max - range.min) * (x ** 2 - x) + range.max;
+
+export function getProgressionColor(scale: number) {
+  const hue = quadradicScale(scale, { min: 200, max: 218 });
+  const saturation = 55;
+  const lightness = Math.round(linearScale(1 - scale, { min: 35, max: 98 }));
+  return `hsl(${hue},${saturation}%,${lightness}%)`;
+}
+
+export function getDataRange(data: Props['data']) {
+  const firstValue = data[0] ? data[0].value : 0;
+  return data.reduce(
+    ([min, max], { value }) => [Math.min(min, value), Math.max(max, value)],
+    [firstValue, firstValue]
+  );
+}
+
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const {
+    style,
+    initialGeojsonSource,
+    geojsonKeyProperty,
+    data,
+    renderTooltip: ToolTip,
+    initialMapboxOptions
+  } = props;
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [map, setMap] = useState<Map | null>(null);
+  const popupRef = useRef<Popup | null>(null);
+  const popupContainerRef = useRef<HTMLDivElement>(null);
+  const enableScrollZoom = useRef(false);
+  const [hoverState, setHoverState] = useState<{
+    geojsonProperties?: GeoJsonProperties;
+    data?: ChoroplethDataElement;
+  }>({});
+  // props stored in useRefs, used in the initialization side effect
+  const initialGeojsonSourceRef = useRef(initialGeojsonSource);
+  const initialMapboxOptionsRef = useRef(initialMapboxOptions);
+
+  // converts input data to a scaled value between 0 and 1
+  const getValueScale = useCallback(
+    (value: number) => {
+      const [min, max] = getDataRange(data);
```

Comment:
Since `getValueScale` is called within a loop `getDataRange` will be calculated for each element in the `data` array

Reply:
Yeah, I'd probably do something like:
```js
  const [min, max] = useMemo(() => getDataRange(data), [data]);
  const getValueScale = useCallback(
    (value: number) => (value - min) / (max - min),
    [min, max]
  );
```